csharp/src/Ice/ConnectionI.cs

```diff
@@ -1655,31 +1643,42 @@ private void setState(int state)
                 case StateActive:
                 {
                     //
-                    // Can only switch from holding or not validated to
-                    // active.
+                    // Can only switch to active from holding or not validated.
                     //
                     if (_state != StateHolding && _state != StateNotValidated)
                     {
                         return;
                     }
+
+                    if (_maxDispatches > 0 && _dispatchCount >= _maxDispatches)
+                    {
+                        // Don't resume reads if we're at or over the _maxDispatches limit.
+                        return;
+                    }
+
                     _threadPool.register(this, SocketOperation.Read);
+                    _idleTimeoutTransceiver?.enableIdleCheck();
                     break;
                 }
 
                 case StateHolding:
                 {
                     //
-                    // Can only switch from active or not validated to
-                    // holding.
+                    // Can only switch to holding from active or not validated.
                     //
                     if (_state != StateActive && _state != StateNotValidated)
                     {
                         return;
                     }
-                    if (_state == StateActive)
+
+                    if (_maxDispatches > 0 && _dispatchCount >= _maxDispatches)
                     {
-                        _threadPool.unregister(this, SocketOperation.Read);
+                        // Reads are already disabled if the _maxDispatches limit is reached or exceeded.
+                        return;
```

Comment:
Should this be a `break;`? Don't we still want to set `_state` to `StateHolding`?

Reply:
Yes, we want to switch state. Fixed.